### PR TITLE
[DPE-2850][FIX] Secret content duplication AND rolling update safe

### PR DESCRIPTION
**Pull request by juditnovak**
A major bug was discovered. When using secrets, all secret info was also duplicated into the databag.

## Details:

On [line 783](https://github.com/canonical/data-platform-libs/blob/main/lib/charms/data_platform_libs/v0/data_interfaces.py#L783) the `second_chance_as_normal_field` variable was always true. (Should have been set `False` in modification functions, by mistake.)

However since in parallel we had concerns regarding rolling upgrades, I provided a fix that also addresses the rolling upgrades issue.

During development [another bug](https://github.com/canonical/data-platform-libs/issues/108) was discovered. To be fixed in a separate PR. Tests `xfail`-ing until.

### CODE REVIEW

**Comment by juditnovak** [comment]
TODO: Increase LIBPATCH!!!

File: lib/charms/data_platform_libs/v0/data_interfaces.py [hunk 1]
```diff
@@ -763,14 +763,23 @@ def _process_secret_fields(
...
         """Isolate target secret fields of manipulation, and execute requested operation by Secret Group."""
         result = {}
+
+        # If the relation started on a databag, we just stay on the databag
+        # (Rolling upgrades may result in a relation starting on databag, getting secrets enabled on-the-fly)
+        # self.local_app is sufficient to check (ignored if Requires, never has secrets -- works if Provides)
+        fallback_to_databag = (
+            req_secret_fields
+            and self.local_unit.is_leader()
+            and set(req_secret_fields) & set(relation.data[self.local_app])
```
Comment (delgod) [inline lib/charms/data_platform_libs/v0/data_interfaces.py:778]: can we use `intersection` for better readability?
```suggestion
            and set(req_secret_fields).intersection(set(relation.data[self.local_app]))
```

**Review by delgod — APPROVED** [review]

File: lib/charms/data_platform_libs/v0/data_interfaces.py (hunk 1 above)
Comment (juditnovak) [reply lib/charms/data_platform_libs/v0/data_interfaces.py:778]: Sure....
(Tho I think it looks a way better as it is ;-) )

File: lib/charms/data_platform_libs/v0/data_interfaces.py [hunk 2]
```diff
@@ -1013,57 +1024,76 @@ def _add_relation_secret(
...
 
         old_content = secret.get_content()
         new_content = copy.deepcopy(old_content)
         for field in fields:
-            new_content.pop(field)
+            try:
+                new_content.pop(field)
+            except KeyError:
+                logging.error(
+                    "Non-exsiting secret was attempted to be removed %s, %s",
```
Comment (welpaolo) [inline lib/charms/data_platform_libs/v0/data_interfaces.py:1081]: typo: `Non-existing`

File: tests/integration/application-charm/src/charm.py [hunk 3]
```diff
@@ -141,27 +157,71 @@ def __init__(self, *args):
...
+        self._relation_endpoints = [
+            self.first_database,
+            self.second_database,
+            self.database_clusters,
+            self.aliased_database_clusters,
+            self.kafka,
+            self.opensearch,
+        ]
+
+    def _get_relation(self, relation_id: int) -> Tuple[CharmBase, Relation]:
```
Comment (welpaolo) [inline tests/integration/application-charm/src/charm.py:179]: Please add docstring here

Comment (welpaolo) [inline tests/integration/application-charm/src/charm.py:179]: Does the return type hint should be `Optional[Tuple[CharmBase, Relation]]` ?

File: tests/integration/test_rolling_upgrade.py [hunk 4]
```diff
@@ -0,0 +1,552 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
```
Comment (welpaolo) [inline tests/integration/test_rolling_upgrade.py:2]: Please update year to 2023 :)

File: tests/integration/test_rolling_upgrade.py [hunk 5]
```diff
@@ -0,0 +1,552 @@
...
+    for unit in ops_test.model.applications[app_name].units:
+        unit_name_with_dash = unit.name.replace("/", "-")
+        complete_command = (
+            "scp tests/integration/data/data_interfaces.py "
+            f"{unit.name}:/var/lib/juju/agents/unit-{unit_name_with_dash}/charm/lib/charms/data_platform_libs/v0/"
+        )
+        x, stdout, y = await ops_test.juju(*complete_command.split())
+
+
+async def upgrade_to_secrets(ops_test, app_name):
```
Comment (welpaolo) [inline tests/integration/test_rolling_upgrade.py:46]: Please docstring

File: tests/integration/test_rolling_upgrade.py [hunk 6]
```diff
@@ -0,0 +1,552 @@
...
+
+SECRET_REF_PREFIX = "secret-"
+
+
+# Global variables to store the relation ID after creation
+# def pytest_namespace():
+#     return {"first_database_relation": None, "second_database_relation": None}
+
+
+async def downgrade_to_databag(ops_test, app_name):
```
Comment (welpaolo) [inline tests/integration/test_rolling_upgrade.py:36]: Please docstring

File: tox.ini [hunk 7]
```diff
@@ -86,6 +86,18 @@ deps =
 commands =
     pytest -v --tb native --log-cli-level=INFO -s {posargs} {[vars]tests_path}/integration/test_charm.py
 
+[testenv:integration-upgrades]
+description = Run database integration tests
+deps =
+    psycopg2-binary
+    pytest
+    juju{env:LIBJUJU_VERSION_SPECIFIER:==2.9.42.4}
```
Comment (welpaolo) [inline tox.ini:94]: Do we want to run those tests only on Juju 2.9 ? Is this done on purpose ?

**Review by welpaolo — APPROVED** [review]
Great work Judit! I really appreciate the work you have done on this PR. 
Just left some minor comments

File: tox.ini (hunk 7 above)
Comment (juditnovak) [reply tox.ini:94]: Nononon, they're running on Juju3 (that's where secrets could make sense at all).
We could extend it for Juju2 as well... Just pls let me do it in another PR... ;-)

The above assignment is to add a default value. But the pipeline is taking it from the matrix, where we enforcece Juju3

**Review by welpaolo — APPROVED** [review]

**Review by marceloneppel — APPROVED** [review]

**Pull request merged**